spanner/metrics.go

```diff
@@ -0,0 +1,474 @@
+			recordedPerAttempt: true,
+		},
+	}
+
+	// Generates unique client ID in the format go-<random UUID>@<hostname>
+	generateClientUID = func() (string, error) {
+		hostname := "localhost"
+		hostname, err := os.Hostname()
+		if err != nil {
+			return "", err
+		}
+		return uuid.NewString() + "@" + strconv.FormatInt(int64(os.Getpid()), 10) + "@" + hostname, nil
+	}
+
+	exporterOpts = []option.ClientOption{}
+)
+
+type metricInfo struct {
+	additionalAttrs    []string
+	recordedPerAttempt bool
+}
+
+type builtinMetricsTracerFactory struct {
+	enabled             bool
+	isDirectPathEnabled bool
+
+	// To be called on client close
+	shutdown func()
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	operationLatencies metric.Float64Histogram
+	attemptLatencies   metric.Float64Histogram
+	operationCount     metric.Int64Counter
+	attemptCount       metric.Int64Counter
+}
+
+func detectClientLocation(ctx context.Context) string {
+	resource, err := gcp.NewDetector().Detect(ctx)
+	if err != nil {
+		return "global"
+	}
+	for _, attr := range resource.Attributes() {
+		if attr.Key == semconv.CloudRegionKey {
+			return attr.Value.AsString()
+		}
+	}
+	// If region is not found, return global
+	return "global"
+}
+
+func newBuiltinMetricsTracerFactory(ctx context.Context, dbpath string, metricsProvider metric.MeterProvider) (*builtinMetricsTracerFactory, error) {
+	clientUID, err := generateClientUID()
+	if err != nil {
+		log.Printf("built-in metrics: generateClientUID failed: %v. Using empty string in the %v metric atteribute", err, metricLabelKeyClientUID)
+	}
+	project, instance, database, err := parseDatabaseName(dbpath)
+	if err != nil {
+		return nil, err
+	}
+
+	tracerFactory := &builtinMetricsTracerFactory{
+		enabled: false,
+		clientAttributes: []attribute.KeyValue{
+			attribute.String(monitoredResLabelKeyProject, project),
+			attribute.String(monitoredResLabelKeyInstance, instance),
+			attribute.String(metricLabelKeyDatabase, database),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+			// Skipping instance config until we have a way to get it
+			attribute.String(monitoredResLabelKeyInstanceConfig, "unknown"),
+			attribute.String(monitoredResLabelKeyLocation, detectClientLocation(ctx)),
+		},
+		shutdown: func() {},
+	}
+
+	tracerFactory.isDirectPathEnabled = false
+
+	var meterProvider *sdkmetric.MeterProvider
+	if metricsProvider == nil {
+		// Create default meter provider
+		mpOptions, err := builtInMeterProviderOptions(project)
+		if err != nil {
+			return tracerFactory, err
+		}
+		meterProvider = sdkmetric.NewMeterProvider(mpOptions...)
+
+		tracerFactory.enabled = true
+		tracerFactory.shutdown = func() { meterProvider.Shutdown(ctx) }
+	} else {
+		switch metricsProvider.(type) {
+		case noop.MeterProvider:
+			tracerFactory.enabled = false
```

Comment:
why not have `tracerFactory.enabled = false` before the switch?

Reply:
Thanks, moved.